server/src/main/java/org/elasticsearch/index/codec/bloomfilter/BloomFilterPostingsFormat.java

```diff
@@ -0,0 +1,472 @@
+                metaOut.writeVInt(bloomFilters.size());
+                for (BloomFilter field : bloomFilters) {
+                    metaOut.writeVInt(state.fieldInfos.fieldInfo(field.field).number);
+                    metaOut.writeVLong(field.startFilePointer);
+                    metaOut.writeVInt(field.bloomFilterSize);
+                }
+                CodecUtil.writeFooter(metaOut);
+            }
+        }
+    }
+
+    private record BloomFilter(String field, long startFilePointer, int bloomFilterSize) {
+
+    }
+
+    static final class FieldsReader extends FieldsProducer {
+        private final FieldsProducer termsReader;
+        private final Map<String, BloomFilter> bloomFilters;
+        private final IndexInput indexIn;
+
+        FieldsReader(SegmentReadState state) throws IOException {
+            final List<Closeable> toCloses = new ArrayList<>(1);
+            try (
+                ChecksumIndexInput metaIn = state.directory.openChecksumInput(
+                    metaField(state.segmentInfo, state.segmentSuffix),
+                    IOContext.READONCE
+                )
+            ) {
+                CodecUtil.checkIndexHeader(
+                    metaIn,
+                    BLOOM_CODEC_NAME,
+                    VERSION_START,
+                    VERSION_CURRENT,
+                    state.segmentInfo.getId(),
+                    state.segmentSuffix
+                );
+                final PostingsFormat postingsFormat = forName(metaIn.readString());
+                final int numBloomFilters = metaIn.readVInt();
+                bloomFilters = new HashMap<>(numBloomFilters);
+                for (int i = 0; i < numBloomFilters; i++) {
+                    final int fieldNumber = metaIn.readVInt();
+                    final long startFilePointer = metaIn.readVLong();
+                    final int bloomFilterSize = metaIn.readVInt();
+                    assert bloomFilterSize == bloomFilterSize(state.segmentInfo.maxDoc())
+                        : "bloom_filter_size=" + bloomFilterSize + ", max_docs=" + state.segmentInfo.maxDoc();
+                    final String fieldName = state.fieldInfos.fieldInfo(fieldNumber).name;
+                    bloomFilters.put(fieldName, new BloomFilter(fieldName, startFilePointer, bloomFilterSize));
+                }
+                CodecUtil.checkFooter(metaIn);
+                termsReader = postingsFormat.fieldsProducer(state);
+                toCloses.add(this.termsReader);
+                indexIn = state.directory.openInput(indexField(state.segmentInfo, state.segmentSuffix), state.context);
+                toCloses.add(indexIn);
+                CodecUtil.checkIndexHeader(
+                    indexIn,
+                    BLOOM_CODEC_NAME,
+                    VERSION_START,
+                    VERSION_CURRENT,
+                    state.segmentInfo.getId(),
+                    state.segmentSuffix
+                );
```

Comment:
Can you also do a `CodecUtil.retrieveChecksum` on `indexIn` to verify that the checksum has a correct structure?

Reply:
added in https://github.com/elastic/elasticsearch/pull/88409/commits/62aedf97a1f8dfa5d6023eb3fd36ddf76af29c63